### PR TITLE
refactor(managers): implement custom managers api

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
- Add `custom/api.ts` which exposes custom manager modules 
- Add getter fn to `custom/index.ts` which exposes custom managers and their methods
- Replace `manager === 'regex'` in the code with `isCustomManager(manager)`
- Modify existing fns in `manager/index.ts` to look for custom managers separately

## Context
Part of https://github.com/renovatebot/renovate/issues/19066

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by RahulGautamSingh** [comment]
Dafting cause : forgot changes to docs :)

**Comment by RahulGautamSingh** [comment]
@HonkingGoose Do you think changes are required so the docs site doesn't break? I checked and it seems fine to me.

**Comment by HonkingGoose** [comment]
We have a `.pages` file somewhere that has the navigation structure (sidebar) for our published docs. Maybe that needs changes?

**Comment by RahulGautamSingh** [comment]
> We have a `.pages` file somewhere that has the navigation structure (sidebar) for our published docs. Maybe that needs changes?

I guess the [sidebar for managers](https://github.com/renovatebot/renovate/blob/main/docs/usage/modules/manager/.pages) is still a work in progress

**Comment by RahulGautamSingh** [comment]
Apart from the coverage it is good to go.

**Comment by rarkins** [comment]
matchManagers=regex should be migrated to matchManagers=custom

**Comment by RahulGautamSingh** [comment]
> matchManagers=regex should be migrated to matchManagers=custom

Abuot that: https://github.com/renovatebot/renovate/issues/19066#issuecomment-1636703858

**Comment by rarkins** [comment]
https://github.com/renovatebot/renovate/issues/19066#issuecomment-1636732859

**Comment by RahulGautamSingh** [comment]
> [#19066 (comment)](https://github.com/renovatebot/renovate/issues/19066#issuecomment-1636732859)

Will it be okay to implement this by migrating `matchManagers` from `custom` to [ ...getCustomManagerList() ];
Or should I handle it in the `util/package-rules/manager.ts` where the actual matching is done.

**Comment by rarkins** [comment]
Handle as a special case in matchManagers

**Comment by RahulGautamSingh** [comment]
There is an undiscussed change in the PR:

In the extract fn during this loop:
https://github.com/RahulGautamSingh/renovate/blob/031d4e617fec6e67f4af541a92b6a6353e300b83/lib/workers/repository/extract/index.ts#L34-L48

We extract the actual custom manager here and the pass it down. So in all fns after this like `extractPackageFile`, `lookup`, `branchify` and `update` the manager will be `regex` and `jsonata` and not `custom`. Is this okay?

I thought passing down `customType` and using it inplace of manager these functions might onlly increase complexity without adding much benefits.

File: lib/workers/repository/extract/index.ts
Comment (RahulGautamSingh) [inline lib/workers/repository/extract/index.ts:47]: There is an undiscussed change here, because there might be other ways to do it too:

In the extract fn during this loop:
The actual custom manager is extracted here and then passed down. So in all fns after this like `extractPackageFile`, `lookup`, `branchify` and `update` the manager will be `regex` and `jsonata` and not `custom`. Is this okay?

I thought passing down `customType` and using it inplace of manager these functions might only increase complexity without adding much benefits.

File: tools/generate-imports.mjs
Comment (viceice) [inline tools/generate-imports.mjs:152]:
```suggestion
    // get managers list
```

File: lib/workers/repository/extract/index.ts
Comment (viceice) [reply lib/workers/repository/extract/index.ts:47]: wouldn't this change package rule apply logic? 🤔 eg custom no longer applies as match manager after extract.

File: lib/workers/repository/extract/index.spec.ts
Comment (viceice) [inline lib/workers/repository/extract/index.spec.ts:29]:
```suggestion
      delete config.regexManagers; // for coverage
```
isn't this better?

File: lib/util/package-rules/managers.spec.ts
Comment (viceice) [inline lib/util/package-rules/managers.spec.ts:55]:
```suggestion
    it('should return null if matchManagers is undefined', () => {
```

File: lib/util/package-rules/managers.spec.ts
Comment (viceice) [inline lib/util/package-rules/managers.spec.ts:7]:
```suggestion
    it('should match againt custom manager when matchManager=custom', () => {
```

Comment (HonkingGoose) [reply lib/util/package-rules/managers.spec.ts:7]:
```suggestion
    it('should match against custom manager when matchManager=custom', () => {
```

Actually, there are two typos. 😉

**Comment by RahulGautamSingh** [comment]
> wouldn't this change package rule apply logic? 🤔 eg custom no longer applies as match manager after extract.

If manager is a custom manager then we check for `custom` in matchManagers as well. So it will be okay

File: lib/workers/repository/extract/index.ts
Comment (RahulGautamSingh) [reply lib/workers/repository/extract/index.ts:47]: It should be fine because of the way custom managers are handled in matchManagers.
If manager is say `regex` we first check if the matchManagers array has `custom` too, instead checking for `regex` only.

Comment (rarkins) [reply lib/workers/repository/extract/index.ts:47]: I'm not sure that this is the best idea, although maybe it's ok. @secustor wdyt?

Comment (secustor) [reply lib/workers/repository/extract/index.ts:47]: So if I understand this correctly the user can match `custom` and sub variants e.g. `regex` using `matchManager`.
 
This should work on first glance, tough I would rather add all sub variants in the style of `custom.regex` to the list.    
That makes it easier understandable that custom and regex are connected and not totally different managers, also this should remove the need for special handling in extract code too.

Further from an UX perspective that would make it more consistent with `customDatasources`.

File: lib/modules/manager/custom/regex/strategies.ts
Comment (RahulGautamSingh) [reply lib/modules/manager/custom/regex/strategies.ts:26]: Some with codecov. It will be removed before merging in main

File: lib/workers/repository/extract/index.ts
Comment (rarkins) [reply lib/workers/repository/extract/index.ts:47]: Yes I like the custom.regex

Comment (RahulGautamSingh) [reply lib/workers/repository/extract/index.ts:47]:
```ts
matchManagers = ['custom.regex', 'custom.jsonata']  // allowed
matchManagers = ['custom', 'regex', 'jsonata'] // not allowed
```

And do we want the same for `enabledManagers`?

Comment (viceice) [reply lib/workers/repository/extract/index.ts:47]: I like the `custom.*` variant 👍

Comment (RahulGautamSingh) [reply lib/workers/repository/extract/index.ts:47]: Should this be the way things are handled internally as well?
ie. in the packageFiles with  updates or managerConfig we return manager as `custom.regex` instead of `regex`

Comment (viceice) [reply lib/workers/repository/extract/index.ts]: yes, I think it's acceptable.

Comment (rarkins) [reply lib/workers/repository/extract/index.ts]: Sounds fine

File: lib/config/massage.ts
Comment (secustor) [inline lib/config/massage.ts:25]:
```suggestion
    } else if (key === 'enabledManagers' && is.array<string>(val, is.string)) {
      let matchMgrs = val
```

File: lib/config/massage.spec.ts
Comment (secustor) [inline lib/config/massage.spec.ts:43]:
```suggestion
        matchManagers: ['npm', 'custom.regex'],
```
I think `custom` shouldn't be here based on previous discussions 🤔

Comment (RahulGautamSingh) [reply lib/config/massage.spec.ts:43]: Maybe keep it there till another PR comes along which can handle regex?

Comment (secustor) [reply lib/config/massage.spec.ts:43]: I think we should migrate 1:1 to `custom.regex` as we have currently only one custom manager. 

@rarkins @viceice  WDYT?

Comment (RahulGautamSingh) [reply lib/config/massage.spec.ts:43]: The migration is the same as you want:
`regex` -> `custom.regex`

File: lib/config/index.ts
Comment (secustor) [inline lib/config/index.ts:29]: We will need separate ones.

File: lib/config/massage.spec.ts
Comment (secustor) [reply lib/config/massage.spec.ts:43]: > I thought of keeping both till regex matching is implemented.

IMO no necessary as we have only one currently and with `jsonata` a second one in the near future.

**Review by secustor — COMMENTED** [review]
See comments and please also test agains a real repo.

File: lib/config/index.ts
Comment (RahulGautamSingh) [reply lib/config/index.ts:29]: Will they be added like:
```js
"regex": {...}

// OR

"custom.regex": {...}

```

**Comment by RahulGautamSingh** [comment]
Test-run: https://github.com/RahulGautamSingh-testing/regex-managers/pull/6

**Comment by RahulGautamSingh** [comment]
Blocked by: https://github.com/renovatebot/renovate/pull/23715

**Comment by RahulGautamSingh** [comment]
test-run:
https://github.com/RahulGautamSingh-testing/r-m/pull/3

Cache for regex is manager will be invalidated :)

**Review by HonkingGoose — APPROVED** [review]
I'm happy with most of the text used in the new code. I'll let the maintainers review the code itself.

File: lib/modules/manager/metadata.spec.ts
Comment (secustor) [inline lib/modules/manager/metadata.spec.ts:10]: Are you sure that you want to flip this?

Comment (RahulGautamSingh) [reply lib/modules/manager/metadata.spec.ts:10]: Yes, I have planned to add readme for custom manager in a separate PR where we move from `regexManagers -> customManagers`, so this test should not try to validate the custom manager's readme file as it will be undefined.

File: lib/config/index.ts
Comment (viceice) [inline lib/config/index.ts:29]: can the `getManagerList` return all managers? so we don't have to clone e arrays? (perf)

if not create a new static list and function like `getAllManagersList()`

File: lib/modules/manager/custom/index.ts
Comment (viceice) [inline lib/modules/manager/custom/index.ts:26]: I don't like this, do we really need it?

File: lib/modules/manager/custom/index.ts
Comment (viceice) [inline lib/modules/manager/custom/index.ts:5]: I think we can simply export the array, the wrapper function don't make the code better or safer

File: lib/modules/manager/fingerprint.spec.ts
Comment (viceice) [inline lib/modules/manager/fingerprint.spec.ts:8]: use allmanagerslist

File: lib/modules/manager/index.ts
Comment (viceice) [inline lib/modules/manager/index.ts:16]:
```suggestion
const customManagerList = Array.from(custommManagers.keys());
```
simplify 😉

probably inline into array below, then you can skip the array.from and we safe some memory

File: lib/modules/manager/index.ts
Comment (viceice) [inline lib/modules/manager/index.ts:23]: better use the iscustommanager function as conditional

File: lib/modules/manager/index.ts
Comment (viceice) [inline lib/modules/manager/index.ts:30]:
```suggestion
  for (const managerName of allManagersList) {
```

File: lib/modules/manager/index.ts
Comment (viceice) [inline lib/modules/manager/index.ts:32]: maybe add a wrapper function for get manager which can be reused at multiple places?

File: lib/workers/repository/extract/extract-fingerprint-config.spec.ts
Comment (viceice) [inline lib/workers/repository/extract/extract-fingerprint-config.spec.ts:82]: use allmanagerslist

File: lib/workers/repository/extract/extract-fingerprint-config.ts
Comment (viceice) [inline lib/workers/repository/extract/extract-fingerprint-config.ts:67]: please move to module scope and normal function definition

File: lib/workers/repository/extract/index.ts
Comment (viceice) [inline lib/workers/repository/extract/index.ts:33]: move to module scope

File: tools/docs/manager.ts
Comment (viceice) [inline tools/docs/manager.ts:22]: this now causes lowercase `regex` 🤔

I think we need to pass a proper display name instead

File: lib/modules/manager/index.ts
Comment (RahulGautamSingh) [reply lib/modules/manager/index.ts:34]: This is the only file where it is used like this, so adding it won't do much benefit.

File: lib/workers/repository/extract/index.ts
Comment (RahulGautamSingh) [reply lib/workers/repository/extract/index.ts]: Could've actually use the old code as is realized after trying to move it to module scope, thanks

**Comment by RahulGautamSingh** [comment]
TestRun: https://github.com/RahulGautamSingh-testing/r-m

File: lib/modules/manager/custom/index.spec.ts
Comment (secustor) [inline lib/modules/manager/custom/index.spec.ts:5]:
```suggestion
    expect(customManager.customManagerList).toBeArrayOf(expect.toBeString());
```

File: lib/modules/manager/index.spec.ts
Comment (secustor) [inline lib/modules/manager/index.spec.ts:14]:
```suggestion
    // no need to check custom managers as they support all datasources
```

**Review by secustor — DISMISSED** [review]
Approving this with the assumption that `regex` will be reworked to `custom.regex` in another PR.

**Comment by RahulGautamSingh** [comment]
> Approving this with the assumption that `regex` will be reworked to `custom.regex` in another PR.

Yes

File: lib/modules/manager/custom/index.ts
Comment (viceice) [inline lib/modules/manager/custom/index.ts:8]:
```suggestion
  return !!customManagers.has(manager);
```
map lookup is faster (in future)

**Review by viceice — APPROVED** [review]

**Pull request merged**
